src/vm/eventtrace.cpp

```diff
@@ -4217,6 +4217,29 @@ VOID EtwCallbackCommon(
     GCEventLevel level = static_cast<GCEventLevel>(Level);
     GCHeapUtilities::RecordEventStateChange(bIsPublicTraceHandle, keywords, level);
 
+    DOTNET_TRACE_CONTEXT * ctxToUpdate;
+    switch(ProviderIndex)
+    {
+    case DotNETRuntime:
+        ctxToUpdate = &MICROSOFT_WINDOWS_DOTNETRUNTIME_PROVIDER_DOTNET_Context;
+        break;
+    case DotNETRuntimeRundown:
+        ctxToUpdate = &MICROSOFT_WINDOWS_DOTNETRUNTIME_RUNDOWN_PROVIDER_DOTNET_Context;
+        break;
+    case DotNETRuntimePrivate:
+        ctxToUpdate = &MICROSOFT_WINDOWS_DOTNETRUNTIME_PRIVATE_PROVIDER_DOTNET_Context;
+        break;
+    case DotNETRuntimeStress:
+        ctxToUpdate = &MICROSOFT_WINDOWS_DOTNETRUNTIME_STRESS_PROVIDER_DOTNET_Context;
+        break;
+    default:
+        _ASSERTE(!"EtwCallbackCommon was called with invalid context");
+        return;
+    }
+
+    ctxToUpdate->EventpipeProvider.Level = Level;
```

Comment:
This could cause trouble. Because this callback is shared by both ETW and EventPipe Level may be referring to the Level and ETW session provided. For example assume that we connected EventPipe session A that has level verbose. Later we connect ETW session B at level Informational. That session connection will call through this code path with Level=Informational, then this code will store that level. The next time a verbose event needs to be logged EventPipeHelper::IsEnabled(...) will return FALSE and the event will be incorrectly skipped.


In terms of resolving it, a quick fix and a longer term fix come to mind (or you may think of a better option)
1) Quick fix - add an argument to this call to distinguish EventPipe callbacks from ETW callbacks and only set the filters for EventPipe callbacks.
2) Longer term fix - Don't maintain two copies of the keyword and level information (DRY again). One copy of this data is already being stored in the EventPipeEventProvider object and a second copy is being stored in the EVENTPIPE_TRACE_CONTEXT. Ideally we wouldn't have an EVENTPIPE_TRACE_CONTEXT and the DOTNET_TRACE_CONTEXT would contain the EventPipeEventProvider object directly. However to make this work would require some further modifications to the pre-generated code + modifications to EventPipe's memory management for providers because right now it assumes that it owns the memory for them and deletes that memory on shutdown. The EventPipe helper IsEnabled() macro relies on the memory never being deleted.


I'd suggest doing the quick fix now (or something similarly simple), and maybe we'd do the longer term one someday, maybe not.